File: packages/twilio_flutter_video/ios/Classes/Bridge/VideoEventMapper.swift
import Foundation
import TwilioVideo

final class VideoEventMapper {
  func roomConnected(room: Room) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .roomConnected,
      room: mapRoom(room, state: .connected),
      participant: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func roomDisconnected(room: Room, error: Error?) -> VideoNativeEvent {
    let state: VideoRoomState = error == nil ? .disconnected : .disconnectedWithError
    return VideoNativeEvent(
      type: .roomDisconnected,
      room: mapRoom(room, state: state),
      participant: nil,
      errorCode: error == nil ? nil : "sdk_failure",
      errorMessage: error?.localizedDescription
    )
  }

  func roomReconnecting(room: Room) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .roomReconnecting,
      room: mapRoom(room, state: .reconnecting),
      participant: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func roomReconnected(room: Room) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .roomReconnected,
      room: mapRoom(room, state: .connected),
      participant: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func participantConnected(_ participant: RemoteParticipant) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .participantConnected,
      room: nil,
      participant: mapParticipant(participant),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func participantDisconnected(_ participant: RemoteParticipant) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .participantDisconnected,
      room: nil,
      participant: mapParticipant(participant),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func dominantSpeakerChanged(_ participant: RemoteParticipant?) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .dominantSpeakerChanged,
      room: nil,
      participant: participant.map { mapParticipant($0) },
      errorCode: nil,
      errorMessage: nil
    )
  }

  func error(code: String, message: String) -> VideoNativeEvent {
    VideoNativeEvent(
      type: .error,
      room: nil,
      participant: nil,
      errorCode: code,
      errorMessage: message
    )
  }

  private func mapRoom(_ room: Room, state: VideoRoomState) -> VideoRoomDto {
    VideoRoomDto(
      sid: room.sid,
      name: room.name ?? "",
      state: state
    )
  }

  private func mapParticipant(_ participant: RemoteParticipant) -> VideoParticipantDto {
    VideoParticipantDto(
      sid: participant.sid,
      identity: participant.identity
    )
  }
}

File: packages/twilio_flutter_video/android/src/main/kotlin/com/twilioflutter/video/bridge/VideoEventMapper.kt
package com.twilioflutter.video.bridge

import com.twilio.video.RemoteParticipant
import com.twilio.video.Room
import com.twilio.video.TwilioException
import com.twilioflutter.video.pigeon.VideoEventType
import com.twilioflutter.video.pigeon.VideoNativeEvent
import com.twilioflutter.video.pigeon.VideoParticipantDto
import com.twilioflutter.video.pigeon.VideoRoomDto
import com.twilioflutter.video.pigeon.VideoRoomState

class VideoEventMapper {
  fun roomConnected(room: Room): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.ROOM_CONNECTED,
      room = mapRoom(room, VideoRoomState.CONNECTED),
    )

  fun roomDisconnected(room: Room, error: TwilioException?): VideoNativeEvent {
    val state =
      if (error != null) VideoRoomState.DISCONNECTED_WITH_ERROR else VideoRoomState.DISCONNECTED
    return VideoNativeEvent(
      type = VideoEventType.ROOM_DISCONNECTED,
      room = mapRoom(room, state),
      errorCode = error?.let { "sdk_failure" },
      errorMessage = error?.message,
    )
  }

  fun roomReconnecting(room: Room): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.ROOM_RECONNECTING,
      room = mapRoom(room, VideoRoomState.RECONNECTING),
    )

  fun roomReconnected(room: Room): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.ROOM_RECONNECTED,
      room = mapRoom(room, VideoRoomState.CONNECTED),
    )

  fun participantConnected(participant: RemoteParticipant): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.PARTICIPANT_CONNECTED,
      participant = mapParticipant(participant),
    )

  fun participantDisconnected(participant: RemoteParticipant): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.PARTICIPANT_DISCONNECTED,
      participant = mapParticipant(participant),
    )

  fun dominantSpeakerChanged(participant: RemoteParticipant?): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.DOMINANT_SPEAKER_CHANGED,
      participant = participant?.let { mapParticipant(it) },
    )

  fun error(code: String, message: String): VideoNativeEvent =
    VideoNativeEvent(
      type = VideoEventType.ERROR,
      errorCode = code,
      errorMessage = message,
    )

  private fun mapRoom(room: Room, state: VideoRoomState): VideoRoomDto =
    VideoRoomDto(
      sid = room.sid,
      name = room.name ?: "",
      state = state,
    )

  private fun mapParticipant(participant: RemoteParticipant): VideoParticipantDto =
    VideoParticipantDto(
      sid = participant.sid,
      identity = participant.identity,
    )
}
